src/hotspot/share/runtime/stackWatermarkSet.cpp

```diff
@@ -0,0 +1,145 @@
+#include "runtime/thread.hpp"
+#include "utilities/debug.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/preserveException.hpp"
+#include "utilities/vmError.hpp"
+
+StackWatermarks::StackWatermarks() :
+    _head(NULL) {}
+
+StackWatermarks::~StackWatermarks() {
+  StackWatermark* current = _head;
+  while (current != NULL) {
+    StackWatermark* next = current->next();
+    delete current;
+    current = next;
+  }
+}
+
+StackWatermark* StackWatermarkSet::head(JavaThread* jt) {
+  return jt->stack_watermarks()->_head;
+}
+
+void StackWatermarkSet::set_head(JavaThread* jt, StackWatermark* watermark) {
+  jt->stack_watermarks()->_head = watermark;
+}
+
+void StackWatermarkSet::add_watermark(JavaThread* jt, StackWatermark* watermark) {
+  assert(!has_watermark(jt, watermark->kind()), "Two instances of same kind");
+  StackWatermark* prev = head(jt);
+  watermark->set_next(prev);
+  set_head(jt, watermark);
+}
+
+static void verify_poll_context() {
+#ifdef ASSERT
+  Thread* thread = Thread::current();
+  if (thread->is_Java_thread()) {
+    JavaThread* jt = static_cast<JavaThread*>(thread);
+    JavaThreadState state = jt->thread_state();
+    assert(state != _thread_in_native, "unsafe thread state");
+    assert(state != _thread_blocked, "unsafe thread state");
+  } else if (thread->is_VM_thread()) {
+  } else {
+    assert_locked_or_safepoint(Threads_lock);
+  }
+#endif
+}
+
+void StackWatermarkSet::before_unwind(JavaThread* jt) {
+  verify_poll_context();
+  if (!jt->has_last_Java_frame()) {
+    // Sometimes we throw exceptions and use native transitions on threads that
+    // do not have any Java threads. Skip those callsites.
+    return;
+  }
+  for (StackWatermark* current = head(jt); current != NULL; current = current->next()) {
+    current->before_unwind();
+  }
+  SafepointMechanism::update_poll_values(jt);
+}
+
+void StackWatermarkSet::after_unwind(JavaThread* jt) {
+  verify_poll_context();
+  if (!jt->has_last_Java_frame()) {
+    // Sometimes we throw exceptions and use native transitions on threads that
+    // do not have any Java threads. Skip those callsites.
+    return;
+  }
+  for (StackWatermark* current = head(jt); current != NULL; current = current->next()) {
+    current->after_unwind();
+  }
+  SafepointMechanism::update_poll_values(jt);
+}
+
+void StackWatermarkSet::on_iteration(JavaThread* jt, const frame& fr) {
+  if (VMError::is_error_reported()) {
+    // Don't perform barrier when error reporting walks the stack.
+    return;
+  }
+  verify_poll_context();
+  for (StackWatermark* current = head(jt); current != NULL; current = current->next()) {
+    current->on_iteration(fr);
+  }
+}
+
+void StackWatermarkSet::start_processing(JavaThread* jt, StackWatermarkKind kind) {
+  verify_poll_context();
+  assert(!jt->is_terminated(), "Poll after termination is a bug");
+  StackWatermark* watermark = get(jt, kind);
+  if (watermark != NULL) {
+    watermark->start_processing();
+  }
+}
```

Comment:
No update poll values here?

Reply:
Like the previous scenario, start_processing may be called on a different thread than Thread::current. The thread the stack belongs to will always update the poll values when it wakes up from the safepoint. Other threads have no business updating it.